src/tif_model_iterator/tif_kernel_iterator.py

```diff
@@ -40,942 +38,309 @@ class tif_kernel_iterator_generator:
 
+        @return Pandas DataFrame with columns: bands + [rd_x, rd_y]
         """
-
-        # Make sure the blue band is used in the third array element.
-        kernel = self.normalize_min_max(self.data[2])
-        new_threshold = round(
-            (initial_threshold * kernel.mean())
-            / (initial_threshold * initial_mean)
-            * initial_threshold,
-            0,
+        print("Creating Pixel Coordinates")
+        start = timer()
+        z_shape = data.shape[0]
+        x_shape = data.shape[1]
+        y_shape = data.shape[2]
+        x_coordinates = [
+            [left_boundary + x for y in range(0, data.shape[2])]
+            for x in range(0, data.shape[1])
+        ]
+        y_coordinates = [
+            [y for y in range(0, data.shape[2])] for x in range(0, data.shape[1])
+        ]
+        rd_x, rd_y = rasterio.transform.xy(
+            self.dataset.transform, x_coordinates, y_coordinates
         )
-        print(new_threshold)
-        copy_kernel = kernel.copy().copy()
-
-        for x in range(len(kernel)):
-            for y in range(len(kernel[x])):
-                if kernel[x][y] == 0:
-                    copy_kernel[x][y] = 1
-                elif kernel[x][y] <= new_threshold:
-                    if kernel[x][y] > 0:
-                        copy_kernel[x][y] = 2
-
-                else:
-                    copy_kernel[x][y] = 3
+        data = np.append(data, rd_x).reshape([z_shape + 1, x_shape, y_shape])
+        data = np.append(data, rd_y).reshape([z_shape + 2, x_shape, y_shape])
+        data = data.reshape(-1, x_shape * y_shape).transpose()
 
-        percentage = round((copy_kernel == 3).sum() / (copy_kernel == 2).sum(), 4)
+        df = pd.DataFrame(
+            data,
+            columns=["band" + str(band) for band in self.bands] + ["rd_x", "rd_y"],
+        )
 
-        return percentage
+        print(
+            f"Pixel coordinates creation finished in: {str(timer() - start)} second(s)"
+        )
+        return df
 
-    def unfadify_tile_kernel(self, kernel):
+    @staticmethod
+    def _filter_out_empty_pixels(df: pd.DataFrame) -> pd.DataFrame:
         """
-        Unfade a kernel, for example to plot it again.
+        Remove pixels which have all RGB values zero, as these correspond to the tif_file not being filled there.
+        Note: RGB corresponds to bands 1, 2, 3.
 
-        A fading kernel uses weights to give other pixel than the center pixel less weight in the prediction.
+        @param df: DataFrame with at least bands 1,2,3 for columns
 
-        @param kernel: A faded kernel that can be unfaded.
-        @return: A unfaded kernel.
+        @return df: Filtered version of df
         """
-        return kernel / self.fade_kernel
+        # We want to have RGB values != 0 for any point we want to predict on RGB is in bands 1,2,3
+        non_empty_pixel_mask = (
+            df[["band" + str(band) for band in [1, 2, 3]]] != 0
+        ).any(axis="columns")
```

Comment:
why the any?

Reply:
So this mask is basically:

For the first 3 bands (RGB) is any value non-zero for this pixel? Because if they are all zero we have an empty pixel (0,0,0) which means it isn't actually filled in the raster and we don't want to predict on it.